### PR TITLE
Task-2: Components and partials

### PR DESCRIPTION
**Pull request by EvgeniyTsarkov**
The following amendments have been introduced:
- breadcrumps have been added for the website's views

Breadcrumps - 1
Breadcrumps - 2

### CODE REVIEW

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using System.Text;
+
+namespace NorthwindWebsite.Presentation.Views.Shared.Components.Breadcrumps;
+
+public class Breadcrumps : ViewComponent
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs:6]: Breadcrumbs*

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            var pathElements = path.Split("ImageUpload");
+            processedPath.Append(pathElements[0] + "Image Upload");
+        }
+        else
+        {
+            processedPath.Append(path);
+        }
+
+
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs:47]: I suppose one empty row is redundant

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+namespace NorthwindWebsite.Presentation.Views.Shared.Components.Breadcrumps;
+
+public class Breadcrumps : ViewComponent
+{
+    public IViewComponentResult Invoke()
+    {
+        var path = HttpContext.Request.Path.ToString();
+
+        var breadcrumps = new StringBuilder("Home");
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs:12]: Let's specify `Home` as constant and name in kinda `NavigationStartPoint`

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        breadcrumps.Append(processedPath.Replace("/", " > "));
+
+        return View("_Breadcrumps", breadcrumps);
+    }
+
+    private static StringBuilder ProcessSpecialCases(string path)
+    {
+        var processedPath = new StringBuilder(string.Empty);
+
+        if (path.Contains("Handle"))
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs:25]: Handle can be also marked as constant

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                processedPath.Append(path.Replace("Handle", "Create New"));
+            }
+            else
+            {
+                var pathElements = path.Split("Handle");
+                processedPath.Append(pathElements[0] + "Edit");
+            }
+        }
+        else if (path.Contains("ImageUpload"))
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumps/Breadcrumps.cs:37]: `ImageUpload` can be marked as constant as well

**Review by dotnet-fizzyy — COMMENTED** [review]
As far as I see you have a little bit complicated you algorithm: 
You are creating a string from your path, after this you are looking for replacement words (Home, Handle and etc) and after this you are replace your `/` with `>`.
But what do you think about simplified version: you are creating a string from your path, splitting it by `/`, start array/collection iteration are words replacement and after that you are pefroming LINQ Aggregation/string.Join with `>` character?

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    public IViewComponentResult Invoke()
+    {
+        var path = HttpContext.Request.Path.ToString();
+
+        var processedArray = ProcessSpecialCases(path);
+
+        var readyBreadcrumbs = string.Join(" > ", processedArray);
+
+        var breadcrembs = string.Concat(NavigationStartPoint, readyBreadcrumbs);
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs:19]: 1. I guess you may combine previous and this row:
```suggestion
        var breadcrembs = string
                               .Join(" > ", processedArray)
                               .Concat(NavigationStartPoint, readyBreadcrumbs);
```
3. breadcrumbs*

Comment (EvgeniyTsarkov) [reply src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs]: Can't see how this chain is possible, did it like this: NavigationStartPoint + string.Join(" > ", processedPathAsList);
to get rid of another valiable.

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        var breadcrumbs = NavigationStartPoint + string.Join(" > ", processedPathAsList);
+
+        return View("_Breadcrumbs", breadcrumbs);
+    }
+
+    private static List<string> ProcessSpecialCases(string path)
+    {
+        var pathAsArray = path.Split('/');
+
+        var pathAsList = new List<string>();
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs:26]: Well, I woudn't say that name is quite good. Let's rename this and previous property like `originalPathSegments` and `breadcrumbPathSegments`

File: src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            if (node == ImageUploadActionName)
+            {
+                pathAsList.Add("Image Upload");
+                break;
+            }
+
+            pathAsList.Add(node);
+        }
+
+        if (pathAsList.Contains("Home/"))
```
Comment (dotnet-fizzyy) [inline src/NorthwindWebsite/Presentation/Views/Shared/Components/Breadcrumbs/Breadcrumbs.cs:53]: Not quite sure about this condition: how is it possible? You have splitted your path by `/` symbol. Also, you can check for this condition in `foreach` loop and just skip it.

**Review by dotnet-fizzyy — APPROVED** [review]

**Pull request merged**